### PR TITLE
Add torchao mps lowbit ops to llama runner

### PR DESCRIPTION
**Pull request by manuelcandales**
Setup ET: https://pytorch.org/executorch/stable/getting-started-setup

Install llama runner requirements
```
bash examples/models/llama/install_requirements.sh
```

Build ET with MPS ON:
```
cmake -DPYTHON_EXECUTABLE=python \
    -DCMAKE_INSTALL_PREFIX=cmake-out \
    -DEXECUTORCH_ENABLE_LOGGING=1 \
    -DCMAKE_BUILD_TYPE=Release \
    -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON \
    -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
    -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON \
    -DEXECUTORCH_BUILD_MPS=ON \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=ON \
    -Bcmake-out .
cmake --build cmake-out -j16 --target install --config Release
```

Build llama runner with torchao mps ops
```
cmake -DPYTHON_EXECUTABLE=python \
    -DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
    -DCMAKE_INSTALL_PREFIX=cmake-out \
    -DCMAKE_BUILD_TYPE=Release \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=ON \
    -DEXECUTORCH_BUILD_TORCHAO=ON \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
    -Bcmake-out/examples/models/llama \
    examples/models/llama
cmake --build cmake-out/examples/models/llama -j16 --target install --config Release
```

Export model. Note: qmode can be any of `torchao:fpa1w` ... `torchao:fpa7w`, and `group_size` can be any of: 32, 64, 128, 256
```
CMAKE_INSTALL_PREFIX=$PWD/cmake-out python -m examples.models.llama.export_llama \
--checkpoint /path/to/model.pth \
--params  /path/to/params.json \
-kv --use_sdpa_with_kv_cache --disable_dynamic_shape --metadata '{"get_bos_id":128000, "get_eos_ids":[128009, 128001]}' \
-qmode "torchao:fpa4w" --group_size 32 \
--output_name path/to/output.pte
```

Run:
```
cmake-out/examples/models/llama/llama_main --model_path=/path/to/output.pte --tokenizer_path=/path/to/tokenizer.model --prompt="Once upon a time,"
```

### CODE REVIEW

File: examples/models/llama/CMakeLists.txt [hunk 1]
```diff
@@ -38,6 +38,7 @@ cmake_dependent_option(
 )
 
 option(EXECUTORCH_BUILD_TORCHAO "Build the torchao kernels" OFF)
+option(EXECUTORCH_BUILD_TORCHAO_MPS "Build the torchao mps kernels" OFF)
```
Comment (kimishpatel) [inline examples/models/llama/CMakeLists.txt:41]: Can we overload TORCHAO instead of doing TORCHAO_MPS

Comment (kimishpatel) [reply examples/models/llama/CMakeLists.txt:41]: Less flags to worry about

File: examples/models/llama/source_transformation/quantize.py [hunk 2]
```diff
@@ -72,12 +72,37 @@ def quantize(  # noqa C901
...
         # Add quantization mode options here: group size, bit width, etc.
         return WeightOnlyInt8QuantHandler(model).quantized_model()
-    elif qmode.startswith("torchao:"):
+    elif qmode.startswith("torchao:fpa"):
+        pattern = r"torchao:fpa(\d+)w"
+        matches = re.findall(pattern, qmode)
+        assert len(matches) == 1, f"Expected 1 match for pattern but got {len(matches)}"
+        bitwidth = int(matches[0][0])
+        _load_torchao_aten_lib(
+            libname="libtorchao_ops_mps_linear_fp_act_xbit_weight_aten"
```
Comment (kimishpatel) [inline examples/models/llama/source_transformation/quantize.py:81]: I thought you had a nice lib name libtorchao_ops_mps? Why this long name?

File: examples/models/llama/source_transformation/quantize.py [hunk 3]
```diff
@@ -72,12 +72,37 @@ def quantize(  # noqa C901
...
+                    precision=torch.float32,
+                    groupsize=group_size,
+                    bitwidth=bitwidth,
+                )
+                .quantize(model)
+                .to("cpu")
+            )
+
+        if verbose:
+            print("quantized model:", model)
```
Comment (kimishpatel) [inline examples/models/llama/source_transformation/quantize.py:98]: Please using logging

**Review by kimishpatel — APPROVED** [review]
Some nits but largely looks good. Please include the output for llama runner in the summary.

Consider fixing the nits. Particularly I dont think we should add TORCHAO_MPS. Just overload TORCHAO one

File: examples/models/llama/source_transformation/quantize.py (hunk 2 above)
Comment (manuelcandales) [reply examples/models/llama/source_transformation/quantize.py]: done!

File: examples/models/llama/CMakeLists.txt (hunk 1 above)
Comment (manuelcandales) [reply examples/models/llama/CMakeLists.txt]: done!

**Pull request merged**